core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -674,13 +596,9 @@ class KafkaController(val config: KafkaConfig, zkUtils: ZkUtils, val brokerState
    * elector
    */
   def startup() = {
-    inLock(controllerContext.controllerLock) {
-      info("Controller starting up")
-      registerSessionExpirationListener()
-      isRunning = true
-      controllerElector.startup
-      info("Controller startup complete")
-    }
+    controllerEventQueue.put(Startup)
+    controllerEventQueue.put(Elect)
```

Comment:
The issue is that once the Startup event is enqueued, additional events could be added to the queue by ZK event thread. Those new events could in theory show up before the Elect event. So, if there is sth that we really want the ControllerThread to complete at the beginning. The safest thing is probably to do that in the thread.